lib/Sugar.ml

```diff
@@ -319,15 +319,43 @@ module Let_binding = struct
     ; lb_local: bool
     ; lb_loc: Location.t }
 
+  (** The sugaring is only valid for some patterns. We have to distinguish
+      between these cases:
+
+      {[
+        let local_ x : string = "hi" (* Annotation on the expression *)
+
+        let local_ x : 'a -> 'a -> 'a = "hi" (* Annotation on the pattern *)
+
+        let (x : string) = local_ "hi" (* Annotation on the pattern *)
+      ]} *)
```

Comment:
I'm confused.  I took a look at the parsetree produced by the compiler for the first two examples here, and they are the same.  So I suppose there is a difference in the ocamlformat parsers - is there an easy way to get ocamlformat to dump the parsetree so I can understand better? (I don't see a -dparsetree equivalent in the options).

Reply:
I use `dune exec -- tools/printast/printast.exe ./test.ml` to read the AST.

The two examples indeed have the same AST, I fixed the comment in a new commit.

Reply:
Thanks!

I think an unfortunate related case is:
```ocaml
let x : 'a . 'a -> 'a -> 'a = local_ "hi"
```
If I understand correctly, this will currently be rewritten in such a way as to fail the roundtrip test because we are missing the attribute on the pattern.  I'm not sure what the best resolution is - we could check perhaps for the attribute when deciding to sugar?

Also, this comment is very helpful, but I think it could be slightly clearer in a few ways:
- Because it's a comment involving `local_`, my first guess when reading it was that "annotation on the ..." was a reference to how `local_` is parsed, but I see now it's a reference to how the type annotation is parsed.  So perhaps say "type annotation" or "constraint node" to make it more obvious.
- The comment and the name of the function refer to "sugar" but don't say what is being sugared.  One of them should mention it's a question about when we can sugar "local_" in particular.
- I think in the first case the type annotation ends up on both the expression and the pattern, not just the expression.

Reply:
Sorry, one more annoying request: The comment should probably explain how the location comparison in the code is accomplishing your goal (though I haven't fully understood this bit yet).

Reply:
The function name is unclear but it's specifically about "local lets" (whether it can be rewritten into `let local_ x ...`).
I'll improve that and look at the new case.

This code is a hack and not the solution. The solution is to work on the parsers:
- The standard parser should output identical ASTs for pretended-identical syntaxes. This makes the rules of what we can rewrite clearer.
- The extended parser should output different ASTs for the different syntaxes. This avoids the need to guess the current syntax and the possible rewrites.

Reply:
As an example, there's `Pexp_record`. The standard parser output the same AST for many syntaxes (`{ x : t = x }` is `{ x = (x : t} }`) but the extended parser output a different AST in every parser rules. (the rules are unchanged, only their output is different)

The `Sugar` function for record expressions have been removed, instead we explicitly rewrite the AST in `Extended_ast.Parse.normalize_mapper`.
The formatting code for records is straightforward.

Reply:
All this makes sense - but what should we do in the short term about failing on
```ocaml
let x : 'a . 'a -> 'a -> 'a = local_ "hi"
```
If my suggestion to check for the presence of the attribute on the pattern and not sugar in that case works, perhaps you can just add that to this PR, perhaps add a little more text to the comment to explain the code and document that we intend to come back, and then we can merge this.

If that doesn't work and no other easy fix is available, then we can just merge it as is, with a comment documenting the issue.

Or are you proposing we shouldn't merge this at all and wait until someone has time to change the extended AST?

Reply:
(Also, the CI tests are currently failing on this PR.  I think we expect the windows test to fail, but not the others - probably it just needs to be rebased).

Reply:
This PR is already fixing bugs, so it can be merged at any time. I'll look at this case and CI shortly.

Reply:
I fixed the bug you mentioned and did a rebase. I hope the comment is clearer too. 

The fix is mostly to not rewrite into the shorter form when that would mess with the formatting code or change the AST. A more concrete AST will allow to remove the tricky code and to rewrite in more cases.

Reply:
Great - thanks for dealing with all these requests!